utils/staking-miner/src/monitor.rs

```diff
@@ -54,17 +63,63 @@ async fn ensure_no_previous_solution<
 	T: EPM::Config + frame_system::Config<AccountId = AccountId>,
 	B: BlockT,
 >(
-	ext: &mut Ext,
+	client: &WsClient,
+	at: B::Hash,
 	us: &AccountId,
 ) -> Result<(), Error<T>> {
-	use EPM::signed::SignedSubmissions;
-	ext.execute_with(|| {
-		if <SignedSubmissions<T>>::get().iter().any(|ss| &ss.who == us) {
-			Err(Error::AlreadySubmitted)
-		} else {
-			Ok(())
+	let indices_key = storage_value(EPM_MODULE_PREFIX, b"SignedSubmissionIndices");
+
+	let indices: SubmissionIndicesOf<T> = get_storage(client, rpc_params! {indices_key, at})
+		.await
+		.map_err::<Error<T>, _>(Into::into)?
+		.unwrap_or_default();
+
+	for (_score, id) in indices {
+		let key = storage_value_by_key::<Twox64Concat>(
+			EPM_MODULE_PREFIX,
+			b"SignedSubmissionsMap",
+			&id.encode(),
+		);
+
+		if let Some(submission) =
+			get_storage::<SignedSubmissionOf<T>>(client, rpc_params! {key, at})
+				.await
+				.map_err::<Error<T>, _>(Into::into)?
+		{
+			if &submission.who == us {
+				return Err(Error::AlreadySubmitted)
+			}
 		}
-	})
+	}
+
+	Ok(())
+}
+
+/// Queries the chain for the best solution and checks whether the computed score
+/// is better than best known.
+async fn ensure_no_better_solution<T: EPM::Config, B: BlockT>(
```

Comment:
This one needs a rewrite, there must have been a misunderstanding about it. 

We want to compare ourselves with two queues.

1. The stuff stored in `SignedSubmissionsMap`. These are unverified solutions, and are sorted based on score. We want to scan these solutions, and then react to them based on `SubmissionStrategy`. Recall that the reason that we might want to do something like `AlwaysSubmit` is that these solutions are unverified, and they might have a false high claimed score. 

2. Then you have `QueuedSolution`. This is a single verified solution, that is the current best solution. In practice, we don't need to check this, since it is always empty during the signed phase (it is only set at the end of the signed phase). But since you already have the code, we should just basically treat it always as `OnlySubmitIfLeading`. As in, we check this, if it exists and if it has a better score than us, then we don't submit. Otherwise, we do.

Reply:
> But since you already have the code, we should just basically treat it always as OnlySubmitIfLeading. As in, we check this, if it exists and if it has a better score than us, then we don't submit. Otherwise, we do.

Not sure if I am reading this correctly, but I think there is no need to check the queued solution - we could instead check the phase. AFAICT in the current code its an invariant that if we set a QueuedSolution, then by the end of the block the phase will not be Signed